datacube_stats/utils/__init__.py

```diff
@@ -75,6 +75,61 @@ def ds_all_float(ds):
     return True
 
 
+def da_nodata(da, default=None):
+    from numpy import nan
```

Comment:
I like them because:

- Makes refactoring easier: you can move function anywhere and things will work. 
- Doesn't pollute module with imported symbols, I think smaller scope -- better scope

Negatives are:

- delays import errors to use-time instead of import time (which could also be a plus, rarely)
- More imports for common stuff
- Import delays happen later (could be a good thing also, example `my-app --help`)

Reply:
In this specific case I wouldn't want `nan` being in the module scope, and the alternative is to use `numpy.nan` in the code which is also a bit ugh.

Reply:
I saw you working on fixing tests in the other branch, so used that as an excuse not to add them just yet, so as not to make this merge even harder for you.

Do you want to create an issue: *add integration tests for iterative mode*?

Reply:
It's recommended against by PEP-8, but I couldn't find the reasoning. I would suspect so that import failures are hit early, and perhaps it's slightly more efficient.

I don't mind at all in this case, and in general, but it is a change of convention to most of the codebases we have.

I did have issues with a test where I wanted to monkey patch `GridWorkflow`, and it was impossible due to being imported within a function, instead of at the modules level. I'm not a fan of monkey patching either, but some of our code is incredibly painful to test without it.

Reply:
We can make this compulsory then, I would use `numpy.nan` in that case.

Reply:
Cool, I think that would be best. :)

I do most of my coding and refactoring in PyCharm, and it's got great support for moving functions and classes between modules, and updating all the imports, so I don't even think about it. Doesn't work between notebooks...